far/stddlg.hpp

```diff
@@ -121,10 +120,11 @@ int GetSearchReplaceString(
 	string& ReplaceStr,
 	string_view TextHistoryName,
 	string_view ReplaceHistoryName,
-	search_case_fold* pCaseFold,
+	bool* pCaseSensitive,
 	bool* pWholeWords,
 	bool* pReverse,
 	bool* pRegexp,
+	bool* pFuzzy,
```

Comment:
I like the idea of this PR, but not quite sure about reintroducing bools all over the place: functions with half a dozen boolean parameters is a pain to support.

Maybe instead:
- Add `fuzzy_icase` to the `search_case_fold` enum
- Add a helper function to convert UI checkboxes state into `search_case_fold`, e.g.:
```C++
search_case_fold ui_to_search_mode(bool IsIcase, bool IsFuzzy)
{
	return IsIcase?
		IsFuzzy? search_case_fold::fuzzy_icase : search_case_fold::icase :
		IsFuzzy? search_case_fold::fuzzy : search_case_fold::exact;
}
```
?

This would probably make the PR smaller as well.

Reply:
I think combining two (out of six) independent Boolean parameters in an enumeration, even if it covers all possible combinations, is still cumbersome. I defined dedicated structs to hold all search options, which nicely enables the use of [Designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers). Please see if you like it.

Reply:
All checks passed, I did not bother looking deeper; my bad. I applied the fix.

BTW, how to find gcc 8.1 build starting from PR? It looks like "continuous-integration/appveyor/pr" check points to Clang build. Do I miss something?

At a higher level:
> [E]ventually it all boils down to a flat list of searchers...

I was more concerned about `GetSearchReplaceString`. A function taking multiple positional parameters of the same type, especially sequentially, is a really bad thing. A dedicated struct combined with designated initializers is a nice way to imitate named parameters available in other languages. A bit funny-looking, but we are talking C++ after all.

Unrelated: Which build is official (published at https://www.farmanager.com/), MSVC, gcc, or Clang? Just curious.

Reply:
> how to find gcc 8.1 build starting from PR?

See the buttons in the "AppVeyor" column on the [main page](https://github.com/FarGroup/FarManager).

> points to Clang build

Yes, it looks like the GitHub check does not support multiple AppVeyor instances and gets overwritten by the last one of them unfortunately.

> Which build is official

MSVC 2019, usually slightly outdated (19.29.30136 currently). You can find that in `far:about`.

Reply:
Got it! Thank you.

Reply:
What's the point of supporting old compilers / language versions?

Reply:
> What's the point of supporting old compilers / language versions?

My personal opinion:
- There is no point. This is an end user project, not a library, a typical user takes prebuilt binaries and does not need to build it in some arcane environment, so we can just require the latest and be done with it.

Project maintainer opinion:
- We need to take into account that newest compilers often have more bugs than features
- There are multiple contributors to the project, supposedly not all of them have the latest or are able/willing to upgrade
- Someone has to reconfigure all the CI workflows to use newer compilers, possibly switching to newer VM images, possibly dealing with issues etc., it takes time and effort
- Since it requires some work, it should bring some visible benefits, e.g. at least make the code significantly simpler in multiple places, remove old workarounds etc.

Given the above, the usual primary targets are:
- MSVC: the **previous** major release (e.g. 2019) + the latest updates.
  If possible, we don't intentionally break the support for (previous - 1) and try to stay compatible with it for a while if people ask so. E.g. the last year it was still possible to compile Far in VS2017, but now it's broken and no one complains about it, so it's time to drop it for good.
  The current major release (e.g. 2022) is supported too of course.
  MSVC is used in production, i.e. it produces binaries published here and on farmanager.com
- GCC: the oldest that is more or less feature-compatible with the currently used MSVC.
  Not used in production, mostly for validation, cross checks, ensuring standard compatibility and extra warnings.
- Clang: the latest will do.
  Not used in production, mostly for validation, cross checks, ensuring standard compatibility and extra warnings.

Reply:
There are two different aspects here: eagerly chasing the latest and greatest compiler / language versions, and ruthlessly cutting off old versions as soon as migration to a new one is completed. As you pointed out, regular upgrading requires resources and should be done only when migration offers enough benefits. As much as I enjoy writing code in "modern C++" (whatever it is), I cannot commit to keeping the project at the technology cutting edge, so not asking for anything here.

On the other hand, supporting old versions is just a liability. I think the only reason to maintain compatibility with gcc 8.1 (especially since gcc 12.2.0 runs as part of the regular checks), is catering for contributors who are unable/unwilling to upgrade. (The overhead of build environment reconfiguring is only partially relevant, as most of the effort belongs to the upgrade side.) Frankly, I think that with the latest versions of all three compilers available for free and reasonably easy to install, it would be very beneficial to persuade such contributors to stop hindering the progress of software development industry. 😀